### PR TITLE
Nested records

### PR DESCRIPTION
**Pull request by aswan**
I think this is ready for review.  Flattened field names in ndjson input are not yet working but this patch set is already quite large, I'd like to do that in a follow-up before this gets stale.

### CODE REVIEW

**Comment by aswan** [comment]
Note to reviewers: I've deliberately separated this into 3 commits that can be reviewed (more or less) independently.  Whether they get squashed or not during merging, I don't care, but the review task is likely to be simpler if you look at the commits individually.

File: ast/ast.go [hunk 1]
```diff
@@ -31,8 +34,11 @@ type BooleanExpr interface {
 
 // FieldExpr is the interface implemented by expressions that reference fields.
 type FieldExpr interface {
-	fieldExprNode()
-	Copy() FieldExpr
+	// AsString() generates the original zql source for a FieldExpr.
```
Comment (henridf) [inline ast/ast.go]: I was about to add a comment like "hmm maybe we can do this differently later" but now I don't see it used anywhere. Do we have a need for it coming up?

Comment (henridf) [reply ast/ast.go]: Oh I see now, we do have a need for it. Hmm, could it be moved somewhere else?  Seems a bit specialized. Just wondering and can also be done later.

**Comment by mccanne** [comment]
Very impressive! 

This problem still exists:
```
% go run ./cmd/zq "* | count() by service" logs/*.log
#0:record[service:string,count:count]
0:DNS;1;]
0:SSH;2;]
0:[\x0aDHCP;38;]
0:[\x0aHTTP;61;]
0:[dhcp;13991;]
0:[dns;201423;]
0:[ftp;25;]
0:[ftp-data;13;]
0:[http;40495;]
0:[pop3;6666;]
0:[smtp;26;]
0:[ssh;19;]
0:[ssl;59952;]
0:[ssl,smtp;4;]
0:[xmpp,ssl;54;]
```
This bug is in master too.  The problem is zson.Record.Strings().

File: ast/ast.go (hunk 1 above)
Comment (aswan) [reply ast/ast.go]: Yeah, good point, I'll move this elsewhere.

**Pull request merged**